lib/_http_server.js

```diff
@@ -220,18 +221,22 @@ function writeHead(statusCode, reason, obj) {
   }
   this.statusCode = statusCode;
 
-  var headers;
+  headers = undefined;
   if (this[outHeadersKey]) {
     // Slow-case: when progressive API and header fields are passed.
-    var k;
-    if (obj) {
+    var key;
+    if (typeof obj === 'object' && obj !== null) {
       var keys = Object.keys(obj);
       for (var i = 0; i < keys.length; i++) {
-        k = keys[i];
-        if (k) this.setHeader(k, obj[k]);
+        key = keys[i];
+        if (key)
+          this.setHeader(key, obj[key]);
       }
+    } else if (obj !== undefined) {
+      throw new ERR_INVALID_ARG_TYPE('headers', 'object', obj);
     }
-    if (k === undefined && this._header) {
+
+    if (key === undefined && this._header) {
```

Comment:
@mscdex: I'm not sure exactly what this does? http2-compat doesn't do anything corresponding?